tensorbay/opendataset/nuScenes/loader.py

```diff
@@ -0,0 +1,279 @@
+    dataset.load_catalog(os.path.join(os.path.dirname(__file__), "catalog.json"))
+
+    for subset in os.listdir(root_path):
+        subset_path = os.path.join(root_path, subset)
+        for segment in _generate_segments(subset_path, subset):
+            dataset.add_segment(segment)
+
+    return dataset
+
+
+def _generate_segments(subset_path: str, subset: str) -> Iterator[FusionSegment]:
+    annotation_files = _load_annotation_files(os.path.join(subset_path, subset))
+
+    for scene in annotation_files["scenes"]:
+        segment = FusionSegment(f"{subset}/{scene['name']}")
+        segment.description = scene["description"]
+        current_frame = scene["first_sample_token"]
+        while current_frame:
+            frame, lidar_ego_pose_token = _load_frame(
+                current_frame,
+                annotation_files,
+                segment.sensors,
+                subset_path,
+            )
+            if "test" not in subset:
+                frame["LIDAR_TOP"].label.box3d = _load_labels(
+                    current_frame,
+                    segment.sensors["LIDAR_TOP"].extrinsics,
+                    lidar_ego_pose_token,
+                    annotation_files,
+                )
+
+            segment.append(frame)
+            current_frame = annotation_files["samples"][current_frame]["next"]
+        yield segment
+
+
+def _load_annotation_files(info_path: str) -> Dict[str, Any]:
+    annotation_files = {
+        "samples": _load_info(os.path.join(info_path, "sample.json")),
+        "frame_data": _load_sample_info(os.path.join(info_path, "sample_data.json")),
+        "calibrated_sensors": _load_info(os.path.join(info_path, "calibrated_sensor.json")),
+        "ego_poses": _load_info(os.path.join(info_path, "ego_pose.json")),
+        "sensor": _load_info(os.path.join(info_path, "sensor.json")),
+    }
+    with open(os.path.join(info_path, "scene.json"), "r") as file:
+        annotation_files["scenes"] = json.load(file)
+    if "test" not in info_path:
+        annotation_files["sample_annotations"] = _load_sample_info(
+            os.path.join(info_path, "sample_annotation.json"), no_key_frame=True
+        )
+        annotation_files["instance"] = _load_info(os.path.join(info_path, "instance.json"))
+        annotation_files["category"] = _load_info(os.path.join(info_path, "category.json"))
+        annotation_files["attribute"] = _load_info(os.path.join(info_path, "attribute.json"))
+        annotation_files["visibility"] = _load_info(os.path.join(info_path, "visibility.json"))
+
+    return annotation_files
+
+
+def _load_info(filepath: str) -> Dict[str, Any]:
+    with open(filepath, "r") as fp:
+        info = json.load(fp)
+    return {item["token"]: item for item in info}
+
+
+def _load_sample_info(filepath: str, no_key_frame: bool = False) -> Dict[str, List[Any]]:
+    with open(filepath, "r") as fp:
+        info = json.load(fp)
+    info_with_keys = defaultdict(list)
```

Comment:
It may not be a good idea using `defaultdict` here, because the `info_with_keys` will be returned   and be searched outside this function. If `info_with_keys["xxx"]` is called by a `sample_token` which does not exist in `info_with_keys`, `defaultdict` will not raise `KeyError` but return a `list`, which makes the error harder to detect.